File: packages/player-web/src/runtime/asset-catalog-index.ts
import type {
  EncodedChunkRecord,
  ByteRange,
  Edge,
  Port,
  ParsedFrontIndex,
  ProductionRendition,
  State,
  Unit,
  UnitBlobRange
} from "@pixel-point/aval-format";

import {
  RuntimePlaybackError,
  normalizeRuntimeFailure,
  type RuntimeFailureContext
} from "./errors.js";

export interface RuntimeCatalogIdIndex<TValue> {
  readonly size: number;
  get(id: string): Readonly<TValue> | undefined;
  require(id: string): Readonly<TValue>;
  keys(): readonly string[];
  values(): readonly Readonly<TValue>[];
}

export interface RuntimeCatalogPortEntry {
  readonly unit: string;
  readonly port: Readonly<Port>;
}

export interface RuntimeCatalogPortIndex {
  readonly size: number;
  get(unit: string, port: string): Readonly<RuntimeCatalogPortEntry> | undefined;
  require(unit: string, port: string): Readonly<RuntimeCatalogPortEntry>;
  values(): readonly Readonly<RuntimeCatalogPortEntry>[];
}

export interface RuntimeCatalogChunk {
  readonly rendition: string;
  readonly unit: string;
  readonly decodeIndex: number;
  readonly ordinal: number;
  readonly record: Readonly<EncodedChunkRecord>;
  readonly blobKey?: string;
  readonly blobRange?: Readonly<UnitBlobRange>;
  readonly relativeRange?: Readonly<ByteRange>;
  readonly range: Readonly<ByteRange>;
}

export interface RuntimeCatalogChunkIndex {
  readonly size: number;
  get(
    rendition: string,
    unit: string,
    decodeIndex: number
  ): Readonly<RuntimeCatalogChunk> | undefined;
  require(
    rendition: string,
    unit: string,
    decodeIndex: number
  ): Readonly<RuntimeCatalogChunk>;
  values(): readonly Readonly<RuntimeCatalogChunk>[];
}

export interface CatalogMapBuildInput {
  readonly frontIndex: Readonly<ParsedFrontIndex>;
  readonly declaredFileLength: number;
}

export interface CatalogMaps {
  readonly renditions: Map<string, Readonly<ProductionRendition>>;
  readonly units: Map<string, Readonly<Unit>>;
  readonly states: Map<string, Readonly<State>>;
  readonly edges: Map<string, Readonly<Edge>>;
  readonly ports: Map<string, Readonly<RuntimeCatalogPortEntry>>;
  readonly chunks: Map<string, Readonly<RuntimeCatalogChunk>>;
}

export function buildCatalogMaps(
  input: Readonly<CatalogMapBuildInput>
): CatalogMaps {
  if (typeof input !== "object" || input === null) {
    throw indexError("asset catalog map input is invalid");
  }
  const frontIndex = input.frontIndex;
  const byteLength = input.declaredFileLength;
  if (
    typeof frontIndex !== "object" ||
    frontIndex === null ||
    !Number.isSafeInteger(byteLength) ||
    byteLength < 1 ||
    frontIndex.header?.declaredFileLength !== byteLength
  ) {
    throw indexError("asset catalog front-index geometry is invalid");
  }
  const manifest = frontIndex.manifest;
  const renditions = indexById<ProductionRendition>(
    manifest.renditions,
    "rendition"
  );
  const units = indexById<Unit>(manifest.units, "unit");
  const states = indexById<State>(manifest.states, "state");
  const edges = indexById<Edge>(manifest.edges, "edge");
  const ports = new Map<string, Readonly<RuntimeCatalogPortEntry>>();
  const chunks = new Map<string, Readonly<RuntimeCatalogChunk>>();
  const unitBlobs = indexUnitBlobs(frontIndex, byteLength);

  for (const unit of manifest.units) {
    if (unit.kind !== "body") continue;
    for (const port of unit.ports) {
      insertUnique(
        ports,
        portIdentity(unit.id, port.id),
        Object.freeze({ unit: unit.id, port }),
        "validated asset contains a duplicate body port"
      );
    }
  }

  for (const blob of unitBlobs.values()) {
    const rendition = renditions.get(blob.rendition);
    const unit = units.get(blob.unit);
    if (rendition === undefined || unit === undefined) {
      throw indexError("validated asset chunk span relation is missing");
    }
    const blobEnd = checkedCatalogRangeEnd(
      blob.offset,
      blob.length,
      byteLength
    );
    for (let decodeIndex = 0; decodeIndex < blob.chunkCount; decodeIndex += 1) {
      const ordinal = blob.chunkStart + decodeIndex;
      const record = frontIndex.records[ordinal];
      if (record === undefined) {
        throw indexError("validated asset chunk array is sparse");
      }
      const recordEnd = checkedCatalogRangeEnd(
        record.byteOffset,
        record.byteLength,
        byteLength
      );
      if (record.byteOffset < blob.offset || recordEnd > blobEnd) {
        throw indexError("validated asset chunk exceeds its unit blob");
      }
      const range = Object.freeze({
        offset: record.byteOffset,
        length: record.byteLength
      });
      const relativeRange = Object.freeze({
        offset: record.byteOffset - blob.offset,
        length: record.byteLength
      });
      insertUnique(
        chunks,
        chunkIdentity(rendition.id, unit.id, decodeIndex),
        Object.freeze({
          rendition: rendition.id,
          unit: unit.id,
          decodeIndex,
          ordinal,
          record,
          blobKey: runtimeUnitBlobKey(rendition.id, unit.id),
          blobRange: blob,
          relativeRange,
          range
        }),
        "validated asset contains a duplicate encoded-chunk identity"
      );
    }
  }

  return Object.freeze({
    renditions,
    units,
    states,
    edges,
    ports,
    chunks
  });
}

export function runtimeUnitBlobKey(rendition: string, unit: string): string {
  return `unit:${rendition}:${unit}`;
}

function indexUnitBlobs(
  frontIndex: Readonly<ParsedFrontIndex>,
  declaredFileLength: number
): ReadonlyMap<string, Readonly<UnitBlobRange>> {
  const result = new Map<string, Readonly<UnitBlobRange>>();
  for (const blob of frontIndex.unitBlobs) {
    checkedCatalogRangeEnd(blob.offset, blob.length, declaredFileLength);
    if (
      !Number.isSafeInteger(blob.chunkStart) ||
      !Number.isSafeInteger(blob.chunkCount) ||
      blob.chunkStart < 0 ||
      blob.chunkCount < 1 ||
      blob.chunkStart > frontIndex.records.length ||
      blob.chunkCount > frontIndex.records.length - blob.chunkStart
    ) {
      throw indexError("validated unit blob chunk span is invalid");
    }
    insertUnique(
      result,
      unitBlobIdentity(blob.rendition, blob.unit),
      blob,
      "validated asset contains a duplicate unit blob"
    );
  }
  return result;
}

export function createCatalogIdIndex<TValue>(
  label: string,
  map: () => ReadonlyMap<string, Readonly<TValue>>,
  context: (id: string) => Readonly<RuntimeFailureContext>
): RuntimeCatalogIdIndex<TValue> {
  return Object.freeze({
    get size(): number {
      return map().size;
    },
    get(id: string): Readonly<TValue> | undefined {
      return map().get(id);
    },
    require(id: string): Readonly<TValue> {
      const value = map().get(id);
      if (value === undefined) {
        throw indexError(`asset catalog ${label} lookup failed`, context(id));
      }
      return value;
    },
    keys(): readonly string[] {
      return Object.freeze([...map().keys()]);
    },
    values(): readonly Readonly<TValue>[] {
      return Object.freeze([...map().values()]);
    }
  });
}

export function createCatalogPortIndex(
  map: () => ReadonlyMap<string, Readonly<RuntimeCatalogPortEntry>>
): RuntimeCatalogPortIndex {
  return Object.freeze({
    get size(): number {
      return map().size;
    },
    get(
      unit: string,
      port: string
    ): Readonly<RuntimeCatalogPortEntry> | undefined {
      return map().get(portIdentity(unit, port));
    },
    require(unit: string, port: string): Readonly<RuntimeCatalogPortEntry> {
      const value = map().get(portIdentity(unit, port));
      if (value === undefined) {
        throw indexError("asset catalog port lookup failed", {
          unit,
          path: port
        });
      }
      return value;
    },
    values(): readonly Readonly<RuntimeCatalogPortEntry>[] {
      return Object.freeze([...map().values()]);
    }
  });
}

export function createCatalogChunkIndex(
  map: () => ReadonlyMap<string, Readonly<RuntimeCatalogChunk>>
): RuntimeCatalogChunkIndex {
  return Object.freeze({
    get size(): number {
      return map().size;
    },
    get(
      rendition: string,
      unit: string,
      decodeIndex: number
    ): Readonly<RuntimeCatalogChunk> | undefined {
      return map().get(chunkIdentity(rendition, unit, decodeIndex));
    },
    require(
      rendition: string,
      unit: string,
      decodeIndex: number
    ): Readonly<RuntimeCatalogChunk> {
      const value = map().get(chunkIdentity(rendition, unit, decodeIndex));
      if (value === undefined) {
        throw indexError("asset catalog encoded-chunk lookup failed", {
          rendition,
          unit,
          decodeIndex
        });
      }
      return value;
    },
    values(): readonly Readonly<RuntimeCatalogChunk>[] {
      return Object.freeze([...map().values()]);
    }
  });
}

export function checkedCatalogRangeEnd(
  offset: number,
  length: number,
  limit: number
): number {
  if (
    !Number.isSafeInteger(offset) ||
    !Number.isSafeInteger(length) ||
    offset < 0 ||
    length < 1 ||
    offset > limit ||
    length > limit - offset
  ) {
    throw indexError("validated asset byte range is unavailable");
  }
  return offset + length;
}

function indexById<TValue extends { readonly id: string }>(
  values: readonly Readonly<TValue>[],
  label: string
): Map<string, Readonly<TValue>> {
  const map = new Map<string, Readonly<TValue>>();
  for (const value of values) {
    insertUnique(
      map,
      value.id,
      value,
      `validated asset contains a duplicate ${label}`
    );
  }
  return map;
}

function insertUnique<TValue>(
  map: Map<string, TValue>,
  key: string,
  value: TValue,
  message: string
): void {
  if (map.has(key)) throw indexError(message);
  map.set(key, value);
}

function portIdentity(unit: string, port: string): string {
  return `${unit}/${port}`;
}

function unitBlobIdentity(rendition: string, unit: string): string {
  return runtimeUnitBlobKey(rendition, unit);
}

function chunkIdentity(
  rendition: string,
  unit: string,
  decodeIndex: number
): string {
  return `${rendition}/${unit}/${String(decodeIndex)}`;
}

function indexError(
  message: string,
  context: Readonly<RuntimeFailureContext> = {}
): RuntimePlaybackError {
  return new RuntimePlaybackError(
    normalizeRuntimeFailure("invalid-asset", message, context)
  );
}

File: flutter/packages/aval_player/lib/src/asset_catalog_index.dart
/// Immutable catalog lookup maps and indexes over a parsed front index.
///
/// Direct port of `packages/player-web/src/runtime/asset-catalog-index.ts`.
/// TS `ReadonlyMap` → Dart `Map` (returned views are copied via
/// `List.unmodifiable`); the `Pick`-narrowed index interfaces map onto full
/// abstract interfaces. `frontIndex.header?.declaredFileLength` becomes a plain
/// field read because [FormatHeader] is non-nullable in the Dart format port.
/// The generic `indexById<TValue extends {id}>` becomes an id-extractor
/// argument because Dart has no structural "has an `id`" bound.
library;

import 'package:aval_format/aval_format.dart'
    show
        AccessUnitRecord,
        BodyUnitV01,
        ByteRange,
        EdgeV01,
        ParsedFrontIndex,
        PortV01,
        RenditionV01,
        StateV01,
        UnitBlobRange,
        UnitV01;

import 'errors.dart'
    show
        RuntimeFailureCode,
        RuntimeFailureContext,
        RuntimePlaybackError,
        normalizeRuntimeFailure;

/// A require/get lookup index keyed by entity id.
abstract interface class RuntimeCatalogIdIndex<TValue> {
  int get size;
  TValue? get(String id);
  TValue require(String id);
  List<String> keys();
  List<TValue> values();
}

/// One body port plus its owning unit id.
class RuntimeCatalogPortEntry {
  const RuntimeCatalogPortEntry({required this.unit, required this.port});

  final String unit;
  final PortV01 port;
}

/// A require/get lookup index of body ports keyed by `unit/port`.
abstract interface class RuntimeCatalogPortIndex {
  int get size;
  RuntimeCatalogPortEntry? get(String unit, String port);
  RuntimeCatalogPortEntry require(String unit, String port);
  List<RuntimeCatalogPortEntry> values();
}

/// One access unit's byte geometry and identity.
class RuntimeCatalogAccessUnit {
  const RuntimeCatalogAccessUnit({
    required this.rendition,
    required this.unit,
    required this.localFrame,
    required this.ordinal,
    required this.record,
    required this.range,
    this.blobKey,
    this.blobRange,
    this.relativeRange,
  });

  final String rendition;
  final String unit;
  final int localFrame;
  final int ordinal;
  final AccessUnitRecord record;
  final String? blobKey;
  final UnitBlobRange? blobRange;
  final ByteRange? relativeRange;
  final ByteRange range;
}

/// A require/get lookup index of access units keyed by `rendition/unit/frame`.
abstract interface class RuntimeCatalogRecordIndex {
  int get size;
  RuntimeCatalogAccessUnit? get(String rendition, String unit, int localFrame);
  RuntimeCatalogAccessUnit require(
    String rendition,
    String unit,
    int localFrame,
  );
  List<RuntimeCatalogAccessUnit> values();
}

/// Input to [buildCatalogMaps].
class CatalogMapBuildInput {
  const CatalogMapBuildInput({
    required this.frontIndex,
    required this.declaredFileLength,
  });

  final ParsedFrontIndex frontIndex;
  final int declaredFileLength;
}

/// The complete set of built catalog lookup maps.
class CatalogMaps {
  const CatalogMaps({
    required this.renditions,
    required this.units,
    required this.states,
    required this.edges,
    required this.ports,
    required this.records,
  });

  final Map<String, RenditionV01> renditions;
  final Map<String, UnitV01> units;
  final Map<String, StateV01> states;
  final Map<String, EdgeV01> edges;
  final Map<String, RuntimeCatalogPortEntry> ports;
  final Map<String, RuntimeCatalogAccessUnit> records;
}

CatalogMaps buildCatalogMaps(CatalogMapBuildInput input) {
  final frontIndex = input.frontIndex;
  final byteLength = input.declaredFileLength;
  if (byteLength < 1 || frontIndex.header.declaredFileLength != byteLength) {
    throw _indexError('asset catalog front-index geometry is invalid');
  }
  final manifest = frontIndex.manifest;
  final renditions = _indexById<RenditionV01>(
    manifest.renditions,
    (value) => value.id,
    'rendition',
  );
  final units = _indexById<UnitV01>(
    manifest.units,
    (value) => value.id,
    'unit',
  );
  final states = _indexById<StateV01>(
    manifest.states,
    (value) => value.id,
    'state',
  );
  final edges = _indexById<EdgeV01>(
    manifest.edges,
    (value) => value.id,
    'edge',
  );
  final ports = <String, RuntimeCatalogPortEntry>{};
  final records = <String, RuntimeCatalogAccessUnit>{};
  final unitBlobs = _indexUnitBlobs(frontIndex, byteLength);

  for (final unit in manifest.units) {
    if (unit.kind != 'body') continue;
    for (final port in (unit as BodyUnitV01).ports) {
      _insertUnique(
        ports,
        _portIdentity(unit.id, port.id),
        RuntimeCatalogPortEntry(unit: unit.id, port: port),
        'validated asset contains a duplicate body port',
      );
    }
  }

  for (var ordinal = 0; ordinal < frontIndex.records.length; ordinal += 1) {
    final record = frontIndex.records[ordinal];
    if (record.renditionIndex < 0 ||
        record.renditionIndex >= manifest.renditions.length ||
        record.unitIndex < 0 ||
        record.unitIndex >= manifest.units.length) {
      throw _indexError('validated asset record relation is missing');
    }
    final rendition = manifest.renditions[record.renditionIndex];
    final unit = manifest.units[record.unitIndex];
    checkedCatalogRangeEnd(
      record.payloadOffset,
      record.payloadLength,
      byteLength,
    );
    final blob = unitBlobs[_unitBlobIdentity(rendition.id, unit.id)];
    if (blob == null) {
      throw _indexError('validated asset record has no containing unit blob');
    }
    final blobEnd = checkedCatalogRangeEnd(blob.offset, blob.length, byteLength);
    final recordEnd = record.payloadOffset + record.payloadLength;
    if (ordinal < blob.sampleStart ||
        ordinal >= blob.sampleStart + blob.sampleCount ||
        record.payloadOffset < blob.offset ||
        recordEnd > blobEnd) {
      throw _indexError('validated asset record exceeds its unit blob');
    }
    final range = ByteRange(
      offset: record.payloadOffset,
      length: record.payloadLength,
    );
    final relativeRange = ByteRange(
      offset: record.payloadOffset - blob.offset,
      length: record.payloadLength,
    );
    _insertUnique(
      records,
      _recordIdentity(rendition.id, unit.id, record.frameIndex),
      RuntimeCatalogAccessUnit(
        rendition: rendition.id,
        unit: unit.id,
        localFrame: record.frameIndex,
        ordinal: ordinal,
        record: record,
        blobKey: runtimeUnitBlobKey(rendition.id, unit.id),
        blobRange: blob,
        relativeRange: relativeRange,
        range: range,
      ),
      'validated asset contains a duplicate access-unit identity',
    );
  }

  return CatalogMaps(
    renditions: renditions,
    units: units,
    states: states,
    edges: edges,
    ports: ports,
    records: records,
  );
}

String runtimeUnitBlobKey(String rendition, String unit) =>
    'unit:$rendition:$unit';

Map<String, UnitBlobRange> _indexUnitBlobs(
  ParsedFrontIndex frontIndex,
  int declaredFileLength,
) {
  final result = <String, UnitBlobRange>{};
  for (final blob in frontIndex.unitBlobs) {
    checkedCatalogRangeEnd(blob.offset, blob.length, declaredFileLength);
    if (blob.sampleStart < 0 ||
        blob.sampleCount < 1 ||
        blob.sampleStart > frontIndex.records.length ||
        blob.sampleCount > frontIndex.records.length - blob.sampleStart) {
      throw _indexError('validated unit blob sample span is invalid');
    }
    _insertUnique(
      result,
      _unitBlobIdentity(blob.rendition, blob.unit),
      blob,
      'validated asset contains a duplicate unit blob',
    );
  }
  return result;
}

RuntimeCatalogIdIndex<TValue> createCatalogIdIndex<TValue>(
  String label,
  Map<String, TValue> Function() map,
  RuntimeFailureContext Function(String id) context,
) {
  return _CatalogIdIndex<TValue>(label, map, context);
}

RuntimeCatalogPortIndex createCatalogPortIndex(
  Map<String, RuntimeCatalogPortEntry> Function() map,
) {
  return _CatalogPortIndex(map);
}

RuntimeCatalogRecordIndex createCatalogRecordIndex(
  Map<String, RuntimeCatalogAccessUnit> Function() map,
) {
  return _CatalogRecordIndex(map);
}

int checkedCatalogRangeEnd(int offset, int length, int limit) {
  if (offset < 0 || length < 1 || offset > limit || length > limit - offset) {
    throw _indexError('validated asset byte range is unavailable');
  }
  return offset + length;
}

class _CatalogIdIndex<TValue> implements RuntimeCatalogIdIndex<TValue> {
  _CatalogIdIndex(this._label, this._map, this._context);

  final String _label;
  final Map<String, TValue> Function() _map;
  final RuntimeFailureContext Function(String id) _context;

  @override
  int get size => _map().length;

  @override
  TValue? get(String id) => _map()[id];

  @override
  TValue require(String id) {
    final value = _map()[id];
    if (value == null) {
      throw _indexError('asset catalog $_label lookup failed', _context(id));
    }
    return value;
  }

  @override
  List<String> keys() => List<String>.unmodifiable(_map().keys);

  @override
  List<TValue> values() => List<TValue>.unmodifiable(_map().values);
}

class _CatalogPortIndex implements RuntimeCatalogPortIndex {
  _CatalogPortIndex(this._map);

  final Map<String, RuntimeCatalogPortEntry> Function() _map;

  @override
  int get size => _map().length;

  @override
  RuntimeCatalogPortEntry? get(String unit, String port) =>
      _map()[_portIdentity(unit, port)];

  @override
  RuntimeCatalogPortEntry require(String unit, String port) {
    final value = _map()[_portIdentity(unit, port)];
    if (value == null) {
      throw _indexError(
        'asset catalog port lookup failed',
        RuntimeFailureContext(unit: unit, path: port),
      );
    }
    return value;
  }

  @override
  List<RuntimeCatalogPortEntry> values() =>
      List<RuntimeCatalogPortEntry>.unmodifiable(_map().values);
}

class _CatalogRecordIndex implements RuntimeCatalogRecordIndex {
  _CatalogRecordIndex(this._map);

  final Map<String, RuntimeCatalogAccessUnit> Function() _map;

  @override
  int get size => _map().length;

  @override
  RuntimeCatalogAccessUnit? get(String rendition, String unit, int localFrame) =>
      _map()[_recordIdentity(rendition, unit, localFrame)];

  @override
  RuntimeCatalogAccessUnit require(
    String rendition,
    String unit,
    int localFrame,
  ) {
    final value = _map()[_recordIdentity(rendition, unit, localFrame)];
    if (value == null) {
      throw _indexError(
        'asset catalog access-unit lookup failed',
        RuntimeFailureContext(
          rendition: rendition,
          unit: unit,
          localFrame: localFrame,
        ),
      );
    }
    return value;
  }

  @override
  List<RuntimeCatalogAccessUnit> values() =>
      List<RuntimeCatalogAccessUnit>.unmodifiable(_map().values);
}

Map<String, TValue> _indexById<TValue>(
  List<TValue> values,
  String Function(TValue) idOf,
  String label,
) {
  final map = <String, TValue>{};
  for (final value in values) {
    _insertUnique(
      map,
      idOf(value),
      value,
      'validated asset contains a duplicate $label',
    );
  }
  return map;
}

void _insertUnique<TValue>(
  Map<String, TValue> map,
  String key,
  TValue value,
  String message,
) {
  if (map.containsKey(key)) throw _indexError(message);
  map[key] = value;
}

String _portIdentity(String unit, String port) => '$unit/$port';

String _unitBlobIdentity(String rendition, String unit) =>
    runtimeUnitBlobKey(rendition, unit);

String _recordIdentity(String rendition, String unit, int localFrame) =>
    '$rendition/$unit/$localFrame';

RuntimePlaybackError _indexError(
  String message, [
  RuntimeFailureContext context = const RuntimeFailureContext(),
]) {
  return RuntimePlaybackError(
    normalizeRuntimeFailure(RuntimeFailureCode.invalidAsset, message, context),
  );
}
